consumers/campaigner/main.go

```diff
@@ -66,128 +114,149 @@ func (h *MessageHandler) HandleMessage(m *nsq.Message) error {
 			logrus.WithFields(logrus.Fields{
 				"user_id":     msg.UserID,
 				"segment_ids": msg.SegmentIDs,
-			}).WithError(err).Error("Unable to fetch subscribers.")
-			break
+			}).WithError(err).Error("unable to fetch subscribers.")
+			return nil
 		}
 
 		if len(subs) == 0 {
 			break
 		}
+		html, err := mustache.ParseString(template.HTMLPart)
+		// todo retry = true/false?
+		if err != nil {
+			logrus.WithError(err).
+				WithFields(logrus.Fields{
+					"user_id":     campaign.UserID,
+					"campaign_id": campaign.ID,
+					"template_id": template.ID,
+				}).
+				Error("unable to parse html_part")
+			return nil
+		}
+		txt, err := mustache.ParseString(template.SubjectPart)
+		// todo retry = true/false?
+		if err != nil {
+			logrus.WithError(err).
+				WithFields(logrus.Fields{
+					"user_id":     campaign.UserID,
+					"campaign_id": campaign.ID,
+					"template_id": template.ID,
+				}).
+				Error("unable to parse text_part")
+			return nil
+		}
+		sub, err := mustache.ParseString(template.TextPart)
+		// todo retry = true/false?
+		if err != nil {
+			logrus.WithError(err).
+				WithFields(logrus.Fields{
+					"user_id":     campaign.UserID,
+					"campaign_id": campaign.ID,
+					"template_id": template.ID,
+				}).
+				Error("unable to parse subject_part")
+			return nil
+		}
 
-		// SES allows to send 50 emails in a bulk sending operation
-		chunkSize := 50
-		for i := 0; i < len(subs); i += chunkSize {
-			end := i + chunkSize
-			if end > len(subs) {
-				end = len(subs)
+		for _, s := range subs {
+			m, err := s.GetMetadata()
+			if err != nil {
+				logrus.WithError(err).
+					WithField("subscriber", s).
+					Error("Unable to get subscriber metadata.")
+				continue
 			}
-
-			var dest []*ses.BulkEmailDestination
-			for _, s := range subs[i:end] {
-				m, err := s.GetMetadata()
-				if err != nil {
-					logrus.WithError(err).
-						WithField("subscriber", s).
-						Error("Unable to get subscriber metadata.")
-
-					continue
-				}
-
-				if s.Name != "" {
-					m["name"] = s.Name
-				}
-
-				url, err := s.GetUnsubscribeURL(msg.UserUUID)
-				if err != nil {
-					logrus.WithError(err).
-						WithField("subscriber", s).
-						Error("Unable to get unsubscribe url.")
-				} else {
-					m["unsubscribe_url"] = url
-				}
-
-				jsonMeta, err := json.Marshal(m)
-				if err != nil {
-					logrus.WithError(err).
-						WithField("subscriber", s).
-						Error("Unable to marshal metadata to json.")
-
-					continue
-				}
-				d := &ses.BulkEmailDestination{
-					Destination: &ses.Destination{
-						ToAddresses: []*string{aws.String(s.Email)},
-					},
-					ReplacementTemplateData: aws.String(string(jsonMeta)),
-				}
-
-				dest = append(dest, d)
+			// merge sub metadata with default template metadata
+			for k, v := range m {
```

Comment:
you need to append the default data to subscriber metadata this way you will override default data with subscriber's metadata

Reply:
@gudgl Hmm I think this way will write only the missing keys in sub metadata with def metadata?

Reply:
but for the next subscriber you will lose the default metadata